src/ConnectAuth.ts

```diff
@@ -0,0 +1,192 @@
+import { AuthError, AuthErrorCode } from './AuthError';
+import { decodeUnverifiedConnectJwt, verifyConnectJwt } from './Jwt';
+import { RequestReader } from './requestReader';
+import { ConnectJwt, ContextQsh, CredentialsLoader, InstallType } from './types';
+
+interface CommonVerifyArgs<E> {
+  baseUrl: string;
+  requestReader: RequestReader;
+  loadCredentials: CredentialsLoader<E>;
```

Comment:
I really like the idea but I don't think it's possible, at least not 100%.

First, the method that computes a query string hash I think is better kept as a callback because sometimes it might not be invoked at all depending on internal logic.

Apart from that, the idea seems just a matter of adding an additional step. Only renaming things around would give us this:

```
const authData = new ExpressAuthDataReader(req)
const result = await ConnectAuth.verifyInstallation({ authData, ...others })
```

My point is: the internals of the reader are transparent , unless you need to override something.

Anyway, you gave me good ideas for better names. I will push another version and you let me know what you think.